prompter-source.lisp

```diff
@@ -700,48 +665,30 @@ If the `active-attributes' slot is NIL, return all attributes keys."))
                 (apply #'remove-from-seq (attributes-keys-non-default source) value)))))
 
 (export-always 'active-attributes)
-(defgeneric active-attributes (suggestion &key source &allow-other-keys)
-  (:method ((suggestion suggestion)
-            &key (source (error "Source required"))
-            &allow-other-keys)
-    (let ((inactive-keys (set-difference (attributes-keys (attributes suggestion))
-                                         (active-attributes-keys source)
-                                         :test #'string=))
-          (attribute-thread nil))
-      (prog1
-          (mapcar
-           (lambda (attribute)
-             (if (functionp (attribute-value attribute))
-                 (progn
-                   (unless attribute-thread (setf attribute-thread (make-attribute-thread source)))
-                   (calispel:! (attribute-channel source) (list suggestion attribute))
-                   (list (attribute-key attribute) ""))
-                 attribute))
-           (remove-if
-            (lambda (attr)
-              (find (attribute-key attr) inactive-keys :test #'string=))
-            (attributes suggestion)))
-        (when attribute-thread
-          ;; Terminate thread:
-          (calispel:! (attribute-channel source) (list nil nil))))))
-  (:documentation "Return the active attributes of SUGGESTION.
-Active attributes are queried from SOURCE."))
-
-(defun make-attribute-thread (source)
-  "Return a thread that is bound to SOURCE and used to compute its `suggestion' attributes asynchronously.
-Asynchronous attributes have a string-returning function as a value."
-  ;; TODO: Notify when done updating, maybe using `update-notifier'?
-  (run-thread "Prompter attribute thread"
-    (sera:nlet lp ((sugg-attr-pair (calispel:? (attribute-channel source))))
-      (destructuring-bind (sugg attr) sugg-attr-pair
-        ;; Recheck type here to protect against race conditions.
-        (when (functionp (attribute-value attr))
-          (setf (alex:assoc-value (attributes sugg) (attribute-key attr))
-                (list
-                 (handler-case (funcall (attribute-value attr) (value sugg))
-                   (error (c)
-                     (format nil "keyword error: ~a" c)))))
-          (lp (calispel:? (attribute-channel source))))))))
+(define-generic active-attributes ((suggestion suggestion)
+                                   &key (source (error "Source required"))
+                                   &allow-other-keys)
+  "Return the active attributes of SUGGESTION.
+Active attributes are queried from SOURCE."
```

Comment:
Perhaps mention briefly the meaning of active in "active attributes"? The docstring feels a bit circular.